plugins/performance-lab/includes/performance-marks/class-perflab-performance-marks.php

```diff
@@ -0,0 +1,302 @@
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Class controlling Dev Tools Performance Marks.
+ *
+ * Leverages the Dev Tools extensibility API to add custom performance data. See https://developer.chrome.com/docs/devtools/performance/extension.
+ *
+ * @since n.e.x.t
+ */
+class Perflab_Performance_Marks {
+
+	/**
+	 * Map of stored metrics that will be added as Dev Tools Performance marks.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var array<string, array<string, mixed>>
+	 */
+	private $marks = array();
+
+	/**
+	 * Array of all plugins and their data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var array<string, array<string, mixed>>
+	 */
+	private $plugins_data = array();
+
+	/**
+	 * Initialize the class including plugin data.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function __construct() {
+		global $wp_scripts;
+
+		if ( ! function_exists( 'get_plugins' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+		$this->plugins_data = get_plugins();
+	}
+
+	/**
+	 * Add a mark to the list of marks.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string               $mark_slug The slug of the mark.
+	 * @param array<string, mixed> $args      Arguments for the mark.
+	 */
+	public function add_mark( string $mark_slug, array $args ): void {
+		$this->marks[ $mark_slug ] = $args;
+	}
+
+	/**
+	 * Get a mark by its slug.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $mark_slug The slug of the mark.
+	 * @return array<string, mixed>|null The mark, or null if not found.
+	 */
+	public function get_mark( string $mark_slug ): ?array {
+		return $this->marks[ $mark_slug ] ?? null;
+	}
+
+	/**
+	 * Get all of the marks.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array<string, array<string, mixed>> All of the marks.
+	 */
+	public function get_all_marks(): array {
+		return $this->marks;
+	}
+
+	/**
+	 * Send the marks to the Dev Tools Performance panel.
+	 *
+	 * Outputs inline JavaScript that uses the Dev Tools Performance API to add marks to the timeline.
+	 *
+	 * This function should be called in the footer after all data has been collected.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function send_marks(): void {
+		global $wp_scripts;
+
+		// Collect any scripts output directly.
+		remove_action( 'wp_footer', array( perflab_performance_marks(), 'send_marks' ), 999 );
+		$manually_output_scripts = $this->get_manually_output_scripts();
+
+		// Add the manually output scripts to the marks.
+		foreach ( $manually_output_scripts as $script ) {
+
+			$this->add_mark(
+				'script_output::' . $script['slug'],
+				array(
+					'path' => $script['path'],
+					'slug' => $script['slug'],
+					'name' => $script['name'],
+				)
+			);
+		}
+
+		foreach ( $wp_scripts->done as $handle ) {
+			$src = $wp_scripts->registered[ $handle ]->src;
+			if ( false === $src ) {
+				continue;
+			}
+			// Gather the plugin slug, name at relative path.
+			$plugin_data = $this->get_plugin_data_from_src( $src );
```

Comment:
The way that the AMP plugin and Origination project determine which plugin enqueued a script is to wrap each of the action callbacks so that we can capture a list of the enqueued handles before and after an action callback runs, and then any new script handles added can then be attributed to the theme/plugin which is identified as the source for the given function/method via the Reflection API.

In the AMP plugin this is the `AMP_Validation_Callback_Wrapper` invocable class https://github.com/ampproject/amp-wp/blob/develop/includes/validation/class-amp-validation-callback-wrapper.php

Here's how this is used: https://github.com/ampproject/amp-wp/blob/60a655cdad9e3431b47b2dcc3b2e5ddb86ac5c5e/includes/validation/class-amp-validation-manager.php#L1429-L1490

That `wrap_hook_callbacks` method runs [at the `all` action](https://github.com/ampproject/amp-wp/blob/60a655cdad9e3431b47b2dcc3b2e5ddb86ac5c5e/includes/validation/class-amp-validation-manager.php#L665).

Reply:
> we can capture a list of the enqueued handles

does this also capture manually output 3p scripts (eg. analytics) which are arguably the hardest to attribute?

My approach was to wait and run late on wp_footer (since I'm doing my output there anyway), then parse thru the global script object's "done" property which should include any enqueued scripts.

Reply:
Sorry, I was referring to scripts being printed manually and not enqueued. But yes, 3P enqueued scripts  are also attributes to the relevant plugins/themes. By computing a diff of the the registered/enqueued scripts before and after each hook callback invocation we can determine which callback introduced a script, and then we can look up the origin of the callback via the Reflection API to find out the plugin or theme.